repose-aggregator/tests/functional-tests/src/integrationTest/groovy/features/filters/samlpolicy/SamlFlow20Test.groovy

```diff
@@ -0,0 +1,421 @@
+        fakeIdentityV2Service.getGenerateTokenFromSamlResponseCount() == 1
+
+        //and: "the response from the origin service is appropriately updated by the filter"
+        // TODO: what is Flow 2.0 going to do to the response?
+
+        when: "the saml:response received by the origin service is unmarshalled"
+        Response response = samlUtilities.unmarshallResponse(mc.handlings[0].request.body as String)
+
+        then: "there should be two assertions"
+        response.assertions.size() == 2
+
+        and: "the first assertion and the saml:response should indicate that Repose is the Issuer"
+        response.assertions[0].issuer.value == SAML_REPOSE_ISSUER
+        response.issuer.value == SAML_REPOSE_ISSUER
+
+        and: "the second Assertion should still have a valid signature and have the original Issuer"
+        response.assertions[1].issuer.value == SAML_EXTERNAL_ISSUER
+        samlUtilities.validateSignature(response.assertions[1].signature)
+
+        and: "the saml:response should have a valid signature since Repose should have signed it"
+        samlUtilities.validateSignature(response.signature)
+
+        where:
+        signatureStatus    | saml
+        "not signed"       | SAML_ONE_ASSERTION_SIGNED
+        "signed"           | SAML_ASSERTION_AND_MESSAGE_SIGNED
+        "signed (invalid)" | SAML_ASSERTION_AND_MESSAGE_SIGNED.replaceFirst("\n", "").replaceFirst("\n", "")
+    }
+
+    def "a saml:response without an assertion should be rejected"() {
+        given: "a saml:response without an assertion"
+        def saml = samlResponse(issuer() >> status())
+
+        when:
+        def mc = deproxy.makeRequest(
+                url: reposeEndpoint + SAML_AUTH_URL,
+                method: HTTP_POST,
+                headers: [(CONTENT_TYPE): CONTENT_TYPE_FORM_URLENCODED],
+                requestBody: asUrlEncodedForm((PARAM_SAML_RESPONSE): encodeBase64(saml)))
+
+        then: "the client gets back a bad response"
+        mc.receivedResponse.code as Integer == SC_BAD_REQUEST
+
+        and: "the request doesn't get to the origin service"
+        mc.handlings.isEmpty()
+    }
+
+    @FailsWith(ConditionNotSatisfiedError)
+    def "a saml:response with an unsigned assertion should be rejected"() {
+        given: "a saml:response with an unsigned assertion"
+        def saml = samlResponse(issuer() >> status() >> assertion([:]))
+
+        when:
+        def mc = deproxy.makeRequest(
+                url: reposeEndpoint + SAML_AUTH_URL,
+                method: HTTP_POST,
+                headers: [(CONTENT_TYPE): CONTENT_TYPE_FORM_URLENCODED],
+                requestBody: asUrlEncodedForm((PARAM_SAML_RESPONSE): encodeBase64(saml)))
+
+        then: "the client gets back a bad response"
+        mc.receivedResponse.code as Integer == SC_BAD_REQUEST
+        mc.receivedResponse.body as String == "All assertions must be signed"
+        mc.receivedResponse.headers.getFirstValue(CONTENT_TYPE) == CONTENT_TYPE_TEXT
```

Comment:
Yeah....  I can remove.  I put it in a few other spots; I'll remove it from those spots, too.